discovery/xds/client_test.go

```diff
@@ -0,0 +1,163 @@
+}
+
+func TestMakeXDSResourceHttpEndpointEmptyServerURLHost(t *testing.T) {
+	endpointURL, err := makeXDSResourceHTTPEndpointURL(ProtocolV3, urlMustParse("grpc://127.0.0.1"), "monitoring")
+
+	require.Empty(t, endpointURL)
+	require.NotNil(t, err)
+	require.Contains(t, err.Error(), "must be either 'http' or 'https'")
+}
+
+func TestMakeXDSResourceHttpEndpoint(t *testing.T) {
+	endpointURL, err := makeXDSResourceHTTPEndpointURL(ProtocolV3, urlMustParse("http://127.0.0.1:5000"), "monitoring")
+
+	require.NoError(t, err)
+	require.Equal(t, endpointURL.String(), "http://127.0.0.1:5000/v3/discovery:monitoring")
+}
+
+func TestCreateNewHTTPResourceClient(t *testing.T) {
+	c := &HTTPResourceClientConfig{
+		HTTPClientConfig: conf.HTTP.HTTPClientConfig,
+		Name:             "test",
+		ExtraQueryParams: url.Values{
+			"param1": {"v1"},
+		},
+		Timeout:         1 * time.Minute,
+		ResourceType:    "monitoring",
+		ResourceTypeURL: "type.googleapis.com/envoy.service.discovery.v3.DiscoveryRequest",
+		Server:          "http://127.0.0.1:5000",
+		ClientID:        "client",
+	}
+
+	client, err := NewHTTPResourceClient(c, ProtocolV3)
+
+	require.NoError(t, err)
+
+	require.Equal(t, client.endpoint, "http://127.0.0.1:5000/v3/discovery:monitoring?param1=v1")
+	require.Equal(t, client.client.Timeout, 1*time.Minute)
+
+}
+
+func createTestHTTPResourceClient(t *testing.T, conf *HTTPResourceClientConfig, protocolVersion ProtocolVersion, responder discoveryResponder) (*HTTPResourceClient, func()) {
+	s := createTestHTTPServer(t, func(request *v3.DiscoveryRequest) (*v3.DiscoveryResponse, error) {
+		require.Equal(t, conf.ResourceTypeURL, request.TypeUrl)
+		require.Equal(t, conf.ClientID, request.Node.Id)
+		return responder(request)
+	})
+
+	conf.Server = s.URL
+	client, err := NewHTTPResourceClient(conf, protocolVersion)
+	require.NoError(t, err)
+
+	return client, s.Close
+}
+
+func TestHTTPResourceClientFetchEmptyResponse(t *testing.T) {
+	client, cleanup := createTestHTTPResourceClient(t, httpResourceConf, ProtocolV3, func(request *v3.DiscoveryRequest) (*v3.DiscoveryResponse, error) {
+		return nil, nil
+	})
+	defer cleanup()
+
+	res, err := client.Fetch(context.Background())
+	require.Nil(t, res)
+	require.NoError(t, err)
+}
+
+func TestHTTPResourceClientFetchFullResponse(t *testing.T) {
+	client, cleanup := createTestHTTPResourceClient(t, httpResourceConf, ProtocolV3, func(request *v3.DiscoveryRequest) (*v3.DiscoveryResponse, error) {
+		if request.VersionInfo == "1" {
+			return nil, nil
+		}
+
+		return &v3.DiscoveryResponse{
+			TypeUrl:     request.TypeUrl,
+			VersionInfo: "1",
+			Nonce:       "abc",
+			Resources:   []*anypb.Any{},
+		}, nil
+	})
+	defer cleanup()
+
+	res, err := client.Fetch(context.Background())
+	require.NoError(t, err)
+	require.NotNil(t, res)
+
+	require.Equal(t, client.ResourceTypeURL(), res.TypeUrl)
+	require.Len(t, res.Resources, 0)
+	// should cache the nonce and version info
```

Comment:
Then we should add this as a message to require.Equal:

require.Equal(t, "abc", client.latestNonce, "Nonce not cached")